sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -438,26 +438,42 @@ protected Symbol visitInPredicate(InPredicate node, ExpressionAnalysisContext co
                 return Literal.NULL;
             }
 
-            Set<Object> rightValues = new HashSet<>();
+            InListExpression listExpression = ((InListExpression) node.getValueList());
+
+            if (listExpression.getValues().isEmpty()) {
+                // TODO:check if this is possible
+                return Literal.BOOLEAN_FALSE;
+            }
+
+            Set<Function> comparisons = new HashSet<>(listExpression.getValues().size());
             for (Expression expression : ((InListExpression) node.getValueList()).getValues()) {
                 Symbol right = expression.accept(this, context);
-                Literal rightLiteral;
-                try {
-                    rightLiteral = Literal.convert(right, leftType);
-                    rightValues.add(rightLiteral.value());
-                } catch (IllegalArgumentException | ClassCastException e) {
-                    throw new IllegalArgumentException(
-                            String.format(Locale.ENGLISH, "invalid IN LIST value %s. expected type '%s'",
-                                    SymbolFormatter.format(right),
-                                    leftType.getName()));
+                if (right.valueType() != leftType){
```

Comment:
could it be that we have arrays here? in that case .equals must be used instead of identity check